cmd/daemon.go

```diff
@@ -126,22 +197,269 @@ func processPullRequestEvent(event *github.PullRequestEvent) {
 	log.Println(fmt.Sprintf("Merge status: %v", merged))
 }
 
-// upHandler tries to bring the deployment up. It may have to clone
-// and check for read access.
+// upHandler tries to bring the deployment online
 func upHandler(w http.ResponseWriter, r *http.Request) {
-	// TODD: Check for repo.
-	// If no repo,
-	// 1. clonec
-	// 2. build
-	// 3. run.
-	// If repo,
-	// Check for existing containers.
-	// Check for existing images.
-	http.Error(w, "not implemented", 501)
+	cli, err := client.NewEnvClient()
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	defer cli.Close()
+
+	// Check for existing git repository, clone if no git repository exists.
+	err = checkForGit(projectDirectory)
+	if err != nil {
+		log.Println("No git repository present - cloning from POST event...")
+		auth, err := getGithubKey()
+		if err != nil {
+			http.Error(w, err.Error(), 500)
+			return
+		}
+
+		// Get github URL from up request for cloning
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			http.Error(w, err.Error(), 500)
+			return
+		}
+		defer r.Body.Close()
+		var upReq UpRequest
+		err = json.Unmarshal(body, &upReq)
+		if err != nil {
+			http.Error(w, err.Error(), 500)
+			return
+		}
+
+		// Clone project
+		remoteURL := upReq.Repo
+		log.Println("Attempting to clone " + remoteURL)
+		_, err = git.PlainClone(projectDirectory, false, &git.CloneOptions{
+			URL:  remoteURL,
+			Auth: auth,
+		})
+		if err != nil {
+			removeContents(projectDirectory)
+			http.Error(w, err.Error(), 500)
```

Comment:
This is the most critical error to handle in this request - **user bootstraps their server, then forgets to add the deploy key to github account**. I think that error might pop out somewhere around [here](https://github.com/src-d/go-git/blob/44c364fe3b7b8cdc0f9623afe870d6781a97ebb4/repository.go#L441) in the `PlainClone` call stack.

For MVP, I think the client needs more information on this specific failure case because we expect it so much.

```go
http.Error(w, errors.New('access to project repository rejected; did you forget to add Inertia's deploy key to your repository settings? <deploy_key>'), http.StatusPreconditionFailed)
```

Reply:
As a side note, we should discuss automatically adding this deploy key to the users GitHub repo. It's easy to do with the API - but it requires them entering a GitHub password.

Reply:
Good point